### PR TITLE
Fix: Manual templates

### PR DESCRIPTION
**Pull request by Meldiron**
## What does this PR do?

Ensures manual templates works properly

## Test Plan

- [x] Add E2E test
- [x] Manual QA

CleanShot 2024-08-13 at 12 50 47@2x
CleanShot 2024-08-13 at 12 50 50@2x

Tests pass locally. If I give up on them, just merge this. Tests are problematic in 1.6.x too
CleanShot 2024-08-14 at 16 44 22@2x

## Related PRs and Issues

x

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [x] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -195,7 +195,7 @@ services:
   appwrite-console:
     <<: *x-logging
     container_name: appwrite-console
-    image: appwrite/console:5.0.0-rc14
+    image: appwrite/console:5.0.0-dev1
```
Comment (christyjacob4) [inline docker-compose.yml:198]: what is the dev tag? lets switch back to the rc

File: src/Appwrite/Platform/Workers/Builds.php [hunk 2]
```diff
@@ -207,7 +207,66 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
+                        $tmpTemplateDirectory .= '/';
+                    }
+
+                    $directorySize = $localDevice->getDirectorySize($tmpTemplateDirectory);
+                    $functionsSizeLimit = (int)System::getEnv('_APP_FUNCTIONS_SIZE_LIMIT', '30000000');
+                    if ($directorySize > $functionsSizeLimit) {
+                        throw new \Exception('Repository directory size should be less than ' . number_format($functionsSizeLimit / 1048576, 2) . ' MBs.');
+                    }
+
+                    Console::execute('tar --exclude code.tar.gz -czf ' . $tmpPathFile . ' -C /tmp/builds/' . \escapeshellcmd($buildId) . '-template' . (empty($templateRootDirectory) ? '' : '/' . $templateRootDirectory) . ' .', '', $stdout, $stderr);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:252]:
```suggestion
                    Console::execute('tar --exclude code.tar.gz -czf ' . $tmpPathFile . ' -C /tmp/builds/' . \escapeshellarg($buildId) . '-template' . (empty($templateRootDirectory) ? '' : '/' . $templateRootDirectory) . ' .', '', $stdout, $stderr);
```

File: src/Appwrite/Platform/Workers/Builds.php [hunk 3]
```diff
@@ -263,23 +322,18 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
-                $templateBranch = $template->getAttribute('branch', '');
+                $tempalteVersion = $template->getAttribute('version', '');
 
                 $templateRootDirectory = $template->getAttribute('rootDirectory', '');
                 $templateRootDirectory = \rtrim($templateRootDirectory, '/');
                 $templateRootDirectory = \ltrim($templateRootDirectory, '.');
                 $templateRootDirectory = \ltrim($templateRootDirectory, '/');
 
-                if (!empty($templateRepositoryName) && !empty($templateOwnerName) && !empty($templateBranch)) {
+                if (!empty($templateRepositoryName) && !empty($templateOwnerName) && !empty($tempalteVersion)) {
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:332]: typo

File: src/Appwrite/Platform/Workers/Builds.php [hunk 4]
```diff
@@ -263,23 +322,18 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
                     // Clone template repo
                     $tmpTemplateDirectory = '/tmp/builds/' . \escapeshellcmd($buildId) . '/template';
 
-                    $cloneType = GitHub::CLONE_TYPE_BRANCH;
-                    if(\str_starts_with($templateBranch, '0.1.')) { // Temporary fix for 1.5. In future versions this only support tag names
-                        $cloneType = GitHub::CLONE_TYPE_TAG;
-                    }
-
-                    $gitCloneCommandForTemplate = $github->generateCloneCommand($templateOwnerName, $templateRepositoryName, $templateBranch, $cloneType, $tmpTemplateDirectory, $templateRootDirectory);
+                    $gitCloneCommandForTemplate = $github->generateCloneCommand($templateOwnerName, $templateRepositoryName, $tempalteVersion, GitHub::CLONE_TYPE_TAG, $tmpTemplateDirectory, $templateRootDirectory);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:336]: same typo here

File: src/Appwrite/Platform/Workers/Builds.php [hunk 5]
```diff
@@ -207,7 +207,66 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
+                    $path = $deviceForFunctions->getPath($deployment->getId() . '.' . \pathinfo('code.tar.gz', PATHINFO_EXTENSION));
+                    $result = $localDevice->transfer($tmpPathFile, $path, $deviceForFunctions);
+
+                    if (!$result) {
+                        throw new \Exception("Unable to move file");
+                    }
+
+                    Console::execute('rm -rf ' . $tmpTemplateDirectory, '', $stdout, $stderr);
+
+                    $source = $path;
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:263]: this assignment doesn't seem very useful . lets just use $path to set the attribute in the lines below

File: src/Appwrite/Platform/Workers/Builds.php [hunk 6]
```diff
@@ -207,7 +207,66 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
+                $templateRepositoryName = $template->getAttribute('repositoryName', '');
+                $templateOwnerName = $template->getAttribute('ownerName', '');
+                $tempalteVersion = $template->getAttribute('version', '');
+
+                $templateRootDirectory = $template->getAttribute('rootDirectory', '');
+                $templateRootDirectory = \rtrim($templateRootDirectory, '/');
+                $templateRootDirectory = \ltrim($templateRootDirectory, '.');
+                $templateRootDirectory = \ltrim($templateRootDirectory, '/');
+
+                if (!empty($templateRepositoryName) && !empty($templateOwnerName) && !empty($tempalteVersion)) {
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:222]: I Dont see an else for this... We should throw an exception if this is the case

File: src/Appwrite/Platform/Workers/Builds.php [hunk 7]
```diff
@@ -207,7 +207,66 @@ protected function buildDeployment(Device $deviceForFunctions, Func $queueForFun
...
+                    $localDevice = new Local();
+
+                    if (substr($tmpTemplateDirectory, -1) !== '/') {
+                        $tmpTemplateDirectory .= '/';
+                    }
+
+                    $directorySize = $localDevice->getDirectorySize($tmpTemplateDirectory);
+                    $functionsSizeLimit = (int)System::getEnv('_APP_FUNCTIONS_SIZE_LIMIT', '30000000');
+                    if ($directorySize > $functionsSizeLimit) {
+                        throw new \Exception('Repository directory size should be less than ' . number_format($functionsSizeLimit / 1048576, 2) . ' MBs.');
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Builds.php:249]: 1048576 what exactly is this number?

Comment (Meldiron) [reply src/Appwrite/Platform/Workers/Builds.php:249]: 1024 * 1024. Not a new code, copy&paste from different place.

File: src/Appwrite/Platform/Workers/Builds.php (hunk 6 above)
Comment (Meldiron) [reply src/Appwrite/Platform/Workers/Builds.php:222]: Not every new build has to have template.

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**